src/magicbox/ResultPreviewsManager.ts

```diff
@@ -0,0 +1,214 @@
+  private lastPreviewsQuery: Promise<ISearchResultPreview[]>;
+  private lastSelectedSuggestion: HTMLElement;
+  private root: HTMLElement;
+
+  public get previewsOwner() {
+    return this.lastSelectedSuggestion;
+  }
+
+  public get hasPreviews() {
+    return !!this.suggestionsPreviewContainer;
+  }
+
+  public get focusedPreviewElement() {
+    if (!this.hasPreviews) {
+      return null;
+    }
+    const focusedElement = this.suggestionsPreviewContainer.findClass(this.options.selectedClass)[0];
+    if (!focusedElement || !focusedElement.classList.contains(this.options.previewClass)) {
+      return null;
+    }
+    return focusedElement;
+  }
+
+  public get previewElements() {
+    if (!this.hasPreviews) {
+      return [];
+    }
+    return this.suggestionsPreviewContainer.findClass(this.options.previewClass);
+  }
+
+  private get suggestionsListbox() {
+    return $$($$(this.element).findClass('coveo-magicbox-suggestions')[0]);
+  }
+
+  private get numberOfResultsPerRow() {
+    const previewSelectables = this.suggestionsPreviewContainer.findClass(this.options.previewClass);
+    if (previewSelectables.length === 0) {
+      return 0;
+    }
+    const firstVerticalOffset = previewSelectables[0].offsetTop;
+    const firstIndexOnNextRow = findIndex(previewSelectables, previewSelectable => previewSelectable.offsetTop !== firstVerticalOffset);
+    return firstIndexOnNextRow !== -1 ? firstIndexOnNextRow : previewSelectables.length;
+  }
+
+  constructor(private element: HTMLElement, options: Partial<IResultPreviewsManagerOptions> = {}) {
+    this.options = defaults(options, <IResultPreviewsManagerOptions>{
+      previewHeaderText: l('QuerySuggestPreview'),
+      previewClass: 'coveo-preview-selectable',
+      selectedClass: 'magic-box-selected'
+    });
+    this.root = Component.resolveRoot(element);
+  }
+
+  public async displaySearchResultPreviewsForSuggestion(suggestion: HTMLElement) {
+    if (this.lastSelectedSuggestion !== suggestion) {
+      await this.loadSearchResultPreviews(suggestion);
+    }
+  }
+
+  public getElementInDirection(direction: Direction) {
+    const previewElements = this.previewElements;
+    const focusedIndex = previewElements.indexOf(this.focusedPreviewElement);
+
+    if (focusedIndex === -1) {
+      return null;
+    }
+
+    if (focusedIndex === 0 && direction === Direction.Left) {
+      return null;
+    }
+
+    return previewElements[(focusedIndex + this.getIncrementInDirection(direction)) % previewElements.length];
+  }
+
+  private getIncrementInDirection(direction: Direction) {
+    switch (direction) {
+      case Direction.Left:
+        return -1;
+      case Direction.Right:
+        return 1;
+      case Direction.Up:
+        return -this.numberOfResultsPerRow;
+      case Direction.Down:
+        return this.numberOfResultsPerRow;
+    }
+  }
+
+  private setHasPreviews(shouldHavePreviews: boolean) {
+    if (this.hasPreviews === !!shouldHavePreviews) {
```

Comment:
Since TypeScript doesn't add type-checks, `shouldHavePreviews` could be `null`. Until `!!` was added, this caused issues.